### ISSUE TITLE
Need some help with troubleshooting

### ISSUE THREAD

**Issue by zell-mbc** [d3473r/jitsi-keycloak#29]
I could do with a little bit of help to clear out some missunderstandings
I have Jitsi, Keycloak and jitsi_keycloak working without any errors. Single server environment.

What I am not clear about is what points at what? Or how the information flow is supposed to work.
I think this is how things are supposed to work, but they don't work for me hence I am looking for help:

    1. The reverse proxy (nginx) points incoming traffic to meet.mydomain.de to jitsi (Port:180)
    2. The Jitsi environment variable TOKEN_AUTH_URL=https://auth.mydomain.de/{room} should redirect right away to Keycloak
    3. Authentication takes place and Keycloak redirects to whatever is set up as base URL in Keycloak, I guess meet.mydomain.de again? But this time with a token added to the url?

Now this is not working for me, Jitsi isn't redirecting to Keycloack, I get the standard Jitsi screen and can start a meeting without being asked for user/password. So either I missunderstood how this is supposed to work or I got a problem to deal with right here.
I also don't understand how jitsi_keycloak would get involved in this setup? Either Keycloak or Jitsi need to pull in jitsi_keycloak:3000 at some point, no?

So I tried another flow:

     1. The reverse proxy (nginx) points incoming traffic to meet.mydomain.de to jitsi_keycloak on port 3000
     2. Authentication takes place and Keycloak (Base URL setting) redirects to what?  

I can sucessfully authenticate that way, but get stuck with where to point to afterwards. I tried meet.mydomain.de but this just gives me a blank screen? It also can't be an internal address such as http://127.0.0.1:180 (Jitsi_web) as my browser wont be able to recognize that one?

You can probably tell that I lack some basic understanding in regards how this is supposed to work. Hence I would apreciate some hints so I know where to search.

**Comment by d3473r** [comment]
Hi, sure no problem, i will try to clear things up for you.
The setup should look like this:

- Keycloak on iam.example.com
- Jitsi on meet.example.com
- jitsi-keycloak on auth.meet.example.com

The auth flow should either start from jitsi-keycloak (selecting the room, then redirect to jitsi with the token) or from jitsi (with the `TOKEN_AUTH_URL` set to jitsi-keycloak auth.meet.example.com). Jitsi then will redirect on the "I am the host" prompt to jitsi-keycloak and the selected room will get filled automatically.
On opening jitsi-keycloak (auth.meet.example.com) either directly or via redirect from jitsi you will get redirected to keycloak (iam.example.com) when you are not logged in yet.

I think the first issue that you are having is not enabling auth on the jitsi side.
I'm using jitsi in docker https://github.com/jitsi/docker-jitsi-meet so i only now about the setup there.
Here are the docs for it https://jitsi.github.io/handbook/docs/devops-guide/devops-guide-docker#authentication.
tldr you have to set the auth variables like in the [readme](https://github.com/d3473r/jitsi-keycloak#jitsi), `ENABLE_AUTH=1`, 'AUTH_TYPE=jwt' and so on.
If you are changing the settings after running jitsi with a persistant volume you have to delete it to apply the environment settings. The default dir for the volume is `CONFIG=~/.jitsi-meet-cfg`, so delete the `~/.jitsi-meet-cfg` folder.

I hope this helps with the first issues, tell me if it worked for you.

**Comment by d3473r** [comment]
You are describing the correct behaviour, after logging in you should get redirected to jitsi-keycloak.
If there was a room preselected you will get redirected again to jitsi, if not you can now the select the room you want to join and click the `Open Jitsi` button.

**Comment by zell-mbc** [comment]
> You are describing the correct behaviour, after logging in you should get redirected to jitsi-keycloak.
And that part is working. I tried starting from jitsi-keycloak /without a room) and from Jitsi (with a room), in both cases I can authenticate and get redirected back to jitsi-keycloak (listening on auth.meet.mydomain…). But that's where things stop with a blank screen, in both cases.

![image](https://user-images.githubusercontent.com/29018767/119511145-6e564100-bd72-11eb-8c55-df91a835dfe9.png)

No errors in the docker log either.

**Comment by nickbouwhuis** [comment]
> > You are describing the correct behaviour, after logging in you should get redirected to jitsi-keycloak.
> 
> And that part is working. I tried starting from jitsi-keycloak /without a room) and from Jitsi (with a room), in both cases I can authenticate and get redirected back to jitsi-keycloak (listening on auth.meet.mydomain…). But that's where things stop with a blank screen, in both cases.
> 
> No errors in the docker log either.

What does your browser console say? Getting the same symptoms here. Browser console says 'Authentication failed' for some reason.

**Comment by Lumrenion** [comment]
I had the same problem. For my keycloak instance, I had to set some CORS headers for this to work:
- Access-Control-Allow-Methods: GET,OPTIONS
- Access-Control-Allow-Origin-List: https://auth.meet.example.com
- Access-Control-Allow-Credentials: true
- Access-Control-Allow-Headers: 'Authorization'

If by any chance, you are using traefik as a reverse proxy like I do, here are my docker labels for these settings:
```
- "traefik.http.routers.keycloak.middlewares=ssoCorsHeaders"
- "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowMethods=GET,OPTIONS"
- "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowOriginList=https://aztg.meet.example.com"
- "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowCredentials=true"
- "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowHeaders=Authorization"
```

**Comment by nickbouwhuis** [comment]
> I had the same problem. For my keycloak instance, I had to set some CORS headers for this to work:
> 
> * Access-Control-Allow-Methods: GET,OPTIONS
> * Access-Control-Allow-Origin-List: https://auth.meet.example.com
> * Access-Control-Allow-Credentials: true
> * Access-Control-Allow-Headers: 'Authorization'
> 
> If by any chance, you are using traefik as a reverse proxy like I do, here are my docker labels for these settings:
> 
> ```
> - "traefik.http.routers.keycloak.middlewares=ssoCorsHeaders"
> - "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowMethods=GET,OPTIONS"
> - "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowOriginList=https://aztg.meet.example.com"
> - "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowCredentials=true"
> - "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowHeaders=Authorization"
> ```

Yep. That's exactly what it was. Just had to upgrade Traefik to a more up-to-date version. Thanks

**Comment by mm28ajos** [comment]
> > I had the same problem. For my keycloak instance, I had to set some CORS headers for this to work:
> > 
> > * Access-Control-Allow-Methods: GET,OPTIONS
> > * Access-Control-Allow-Origin-List: https://auth.meet.example.com
> > * Access-Control-Allow-Credentials: true
> > * Access-Control-Allow-Headers: 'Authorization'
> > 
> > If by any chance, you are using traefik as a reverse proxy like I do, here are my docker labels for these settings:
> > ```
> > - "traefik.http.routers.keycloak.middlewares=ssoCorsHeaders"
> > - "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowMethods=GET,OPTIONS"
> > - "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowOriginList=https://aztg.meet.example.com"
> > - "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowCredentials=true"
> > - "traefik.http.middlewares.ssoCorsHeaders.headers.accessControlAllowHeaders=Authorization"
> > ```
> 
> Yep. That's exactly what it was. Just had to upgrade Traefik to a more up-to-date version. Thanks

I am trying to make this work behind a ngnix reverse proxy but failing so far with the same issue (Authorizsation failed after redirect back from keycloak to auth.meet). Did you additionally change anything in your keycloak config?

**Comment by Lumrenion** [comment]
I did not change anything inside keycloak besides what is described in this containers readme. With nginx as reverse proxy I have no experience, but you should be able to set those headers inside your nginx config in the server directive that forwards to your keycloak container.

**Comment by mm28ajos** [comment]
Thanks, I will give it some more tries.

**Comment by d3473r** [comment]
Ahh, sorry guys. I didn't had this issue as i had allowed all origins in my keycloak settings for testing.
You don't have to configure anything in your reverse proxies (nginx, traefik, ...).
Instead you have to allow the origin in your keycloak settings like so:
<img width="359" alt="keycloak" src="https://user-images.githubusercontent.com/10356892/120614594-b0a90d80-c457-11eb-9739-9a232f693e7a.png">

I will add this info to the Readme :)

**Comment by Lumrenion** [comment]
Thank you, I just removed the headers and added this Web Origin. It does not work with https://auth.meet.example.com/* but it works without `/*` for both with and without a room appended to the URL
That is a lot better :)

**Comment by zell-mbc** [comment]
In my case I get an invalid parameter error from Keycloak when I remove the "/*", adding it back takes me to the logon dialog as expect. But then the way back is still the blank screen.

I didn't have an entry in WebOrigins entry before, adding it doesn't seem to make a difference, no matter with or w/o "/*"

It may now be down to my nginx settings, will double check these next but running out of time right now. If anyone who's using nginx and has got a working setup can post his settings in the meantime I would appreciate.

**Comment by mm28ajos** [comment]
@zell-mbc I will also try with ngnix be next days.

**Comment by zell-mbc** [comment]
Had another look into my nginx settings.
Added the 4 lines below, but can't make things work. I am also not sure if I need those settings in the nginx Keycloak section or the jitsi_keycloak one? In any case, I tried both without success.

```
    add_header Access-Control-Allow-Origin "https://auth.meet.mydomain.de";
    add_header Access-Control-Allow-Credentials "true";
    add_header Access-Control-Allow-Methods "GET, OPTIONS";
    add_header Access-Control-Allow-Headers "Authorization";

```
The errors in the browser console are:
  'Authentication failed', 'Failed to load user profile' and 'CORS Allow Origin Not Matching Origin'

**Comment by d3473r** [comment]
You don't need any additional access headers in your reverse proxies. This even degrades your security as keycloak is tricked into thinking the requests are coming from a trusted origin.
If you are getting a cors error you can check the failing request in the network tools and take a look at the `origin` url.
This url has to be added in keycloak to the web origins array. The CORS error will be gone now.

**Comment by mm28ajos** [comment]
For me, the issue was that I created a "confidential" client not a "public" client in keycloak. After changing this and updating the keycloak.json everything was oke.

**Comment by zell-mbc** [comment]
> If you are getting a CORS error you can check the failing request in the network tools and take a look at the `origin` url.
> This url has to be added in keycloak to the web origins array. The CORS error will be gone now.

That was it! I had no idea what to put where, so your hint  allowed me to finally fix this. Boy, this has been quite a journey for me. Really grateful for your help!

So to summarize for other nginx users which may run into this as well:
My keycloak.conf location block:

    location / {
        proxy_pass           http://localhost:7080/;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
     }

My Jitsi-Keycloak location block:
```
   location / {
        proxy_pass           http://localhost:4000/;
    }
```
My Keycloak Client settings:

    Valid redirect URIs: https://auth.meet.mydomain.de/* 
    Web Origins: https://auth.meet.mydomain.de

**Comment by zell-mbc** [comment]
Turns out that things work fine with a browser but the Jitsi Electron App can't connect.
After selecting "I am the host" the screen stays blank and the Electron console shows the errors below:

    WebSocket connection to 'wss://meet.WebSocket connection to 'wss://meet.mydomain.de/xmpp-websocket?room=digitalmentor-mbc' failed: Data frame received after close.de/xmpp-websocket?room=digitalmentor-mbc' failed: Data frame received after close
    Refused to frame 'https://iam.mydomain.de/' because an ancestor violates the following Content Security Policy directive: "frame-ancestors 'self'".
    Refused to frame 'https://iam.mydomain.de/' because an ancestor violates the following Content Security Policy directive: "frame-ancestors iam.mydomain.de auth.meet.mydomain.de meet.mydomain.de".

This surely is just yet another small nginx setting, but I can't figure it out. Have tried various options in nginx without succes I am afraid:

    add_header Content-Security-Policy "frame-ancestors iam.mydomain.de auth.meet.mydomain.de meet.mydomain.de;";

**Comment by mm28ajos** [comment]
Did you check this? https://github.com/matrix-org/prosody-mod-auth-matrix-user-verification

**Comment by zell-mbc** [comment]
> Did you check this? https://github.com/matrix-org/prosody-mod-auth-matrix-user-verification
Hm, not sure how this relates to the final issue?

Just to be clear, the whole authentication process is now working for me when using Firefox and Chromium. It's just the native Jitsi App which somehow produces this iframe (?) issue above. I thought the native Jitsi App is in fact just an Electron wrapper around a Chromium browser which is why it's strange that the JitsiApp is behaving differently to a regular Chromium session.

**Comment by mm28ajos** [comment]
I am also interested in making this work but have not been successful so far...

**Comment by mm28ajos** [comment]
@zell-mbc: Made it working by setting up an additional prosody instance (as not two authentication methods can be used at the same time) with the plugin from [here](https://github.com/matrix-org/prosody-mod-auth-matrix-user-verification) works fine. So the remaining question is actually not liked to this repo and could be closed.

However, the jitsi docker environment currently does not support the auth method against synapse/matrix which makes it difficult to set up (created a pull request to add easy configuration in jitsi, refer to [this](https://github.com/jitsi/docker-jitsi-meet/pull/1084)

**Comment by andrebiegel** [comment]
Hey i do have a similar problem .. My Jitsi-Keycloak  does redirect back to jitsi and in my browser is this CORS problem mentioned .. but i do have the Web Origin adapted.. Does anybody has idea ?

**Comment by mm28ajos** [comment]
@andrebiegel Did you check https://github.com/d3473r/jitsi-keycloak/issues/29#issuecomment-855383949 ?

**Comment by d3473r** [comment]
@andrebiegel can you post a screenshot of the network tab of the failing request please?
Chrome: https://developer.chrome.com/docs/devtools/network/
Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/request_details

**Comment by andrebiegel** [comment]
![grafik](https://user-images.githubusercontent.com/18468602/129526093-22f419b4-1080-44a6-854a-0d1740d73fe0.png)
 the keylcoak client definiton is set on public 
the keycloak version is 14
client config
![grafik](https://user-images.githubusercontent.com/18468602/129527002-35c4370b-8dc3-4c57-a5a4-d489da3a002d.png)
but keycloak seems not to write an Access-Control-Allow-Origin header :-(

**Comment by d3473r** [comment]
@andrebiegel i need to see the failing request in the network tab, especially in the Request Header the origin field and the Response Header

**Comment by martux69** [comment]
Hi,
though reading this thread carefully multiple times I don't really understand where the connection is between the keycloak system and the jitsi-keycloak container ;-( 
In my actual configuration with a running docker jitsi-keycloak container I get directly a grey screen when I call the URL of my selfhosted jitsi system (native installation).
My constellation is:
- jitsi meet is running natively on a public vm at my provider (eg foo-xy.example.com)
- jitsi-keycloak container is also running there (eg. localhost:3000 ?)
- keycloak system is running in my LAN behind a nginx reverse proxy (eg. foo-abc.lan.example.com)
I don't know which is the right value for the tokenAuthUrl ? And what is the value for auth-server-url in the keycloak.json?
In the moment tokenAuthUrl is set to https://localhost:3000/{room} and auth-server-url is set to foo-abc.lan.example.com, but this is not working as said.
Any help is welcome :-)

**Comment by phagemann** [comment]
I am experiencing the same issue as andrebiegel.
Running the following containers all behind traefik:
- Jitsi: meet.DOMAIN.de
- Keycloak: auth.DOMAIN.de
- Jitsi-Keycloak: auth.meet.DOMAIN.de
 
I've included the requested network tab below @d3473r :
![Untitled](https://user-images.githubusercontent.com/18285350/158180277-35b5308f-800e-474a-8401-051ead3f0fcc.png)


Please let me know in case I can supply more/other information.

**Comment by d3473r** [comment]
Hi @phagemann, your error shows `Access-Control-Allow-Origin` missing, that means you haven't add the app's FQDN to the Web Origins in keycloak

**Comment by phagemann** [comment]
@d3473r the FQDN of Jitsi-Keycloak is included in Keycloak, see below. 
(```https://auth.domain.de``` is included from previous test, should be unnecessary)
![Untitled](https://user-images.githubusercontent.com/18285350/158697918-9a97c39a-eca8-408f-ad8b-b8848319c316.png)

When running without a Web Origins entry I get a blank screen after being redirected by Jitsi.
With ```https://auth.meet.MYDOMIN.de``` the Vue app loads, but doesn't forward back to Jitsi, resulting in the error above. 
Filling in a room and clicking the "Open Jitsi" Button throws an ```TypeError: this.config is null```.  

Did I get the instructions wrong?

**Comment by phagemann** [comment]
When setting up jitsi-keycloak on a different server than keycloak and jitsi, it works just fine. Previously they all ran on the same machine behind traefik.
The issue seems to be related to traefik / networking. ATM I don't understand why. 

Never the less this is not an issue caused by jitsi-keycloak. @d3473r  Thank you for your support =)

**Comment by phagemann** [comment]
For anyone stumbling upon the same issue:
As Jitsi-Keycloak, Jitsi and Keycloak all use Traefik as reverseproxy they share a common network.
The docker-internal DNS resolver returns the docker-internal IP of **Keycloak** instead of **Traefik**, which would be mandatory to reach Keycloak via HTTPS.

This could be fixed by assigning Traefik the FQDN of the participant Hosts as alias, where ```traefik_net``` is the common network:

```
services:
  traefik:
    image: ...
    networks:
      traefik_net:
        aliases:
          - "auth.MYDOMAIN.de"
          - "chat.MYDOMAIN.de"
```
